File: account/src/main/java/serp/project/account/ui/controller/AuthController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import serp.project.account.core.domain.dto.request.CreateUserDto;
import serp.project.account.core.domain.dto.request.LoginRequest;
import serp.project.account.core.domain.dto.request.RefreshTokenRequest;
import serp.project.account.core.domain.dto.request.RevokeTokenRequest;
import serp.project.account.core.usecase.AuthUseCase;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/auth")
public class AuthController {
    private final AuthUseCase authUseCase;

    @PostMapping("/register")
    public ResponseEntity<?> register(@Valid @RequestBody CreateUserDto request) {
        var response = authUseCase.registerUser(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/login")
    public ResponseEntity<?> login(@Valid @RequestBody LoginRequest request) {
        var response = authUseCase.login(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/get-token")
    public ResponseEntity<?> getToken(@Valid @RequestBody LoginRequest request) {
        var response = authUseCase.getUserToken(request.getEmail(), request.getPassword());
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/refresh-token")
    public ResponseEntity<?> refreshToken(@Valid @RequestBody RefreshTokenRequest request) {
        var response = authUseCase.refreshToken(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/revoke-token")
    public ResponseEntity<?> revokeToken(@Valid @RequestBody RevokeTokenRequest request) {
        var response = authUseCase.revokeToken(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }
}

File: api_gateway/src/ui/controller/account/auth_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type AuthController struct {
	authService service.IAuthService
}

func (a *AuthController) Register(c *gin.Context) {
	var req request.RegisterDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := a.authService.Register(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}

	c.JSON(res.Code, res)
}

func (a *AuthController) Login(c *gin.Context) {
	var req request.LoginDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := a.authService.Login(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}

	c.JSON(res.Code, res)
}

func (a *AuthController) RefreshToken(c *gin.Context) {
	var req request.RefreshTokenDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := a.authService.RefreshToken(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (a *AuthController) GetToken(c *gin.Context) {
	var req request.LoginDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := a.authService.GetToken(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}

	c.JSON(res.Code, res)
}

func (a *AuthController) RevokeToken(c *gin.Context) {
	var req request.RefreshTokenDTO
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := a.authService.RevokeToken(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewAuthController(authService service.IAuthService) *AuthController {
	return &AuthController{
		authService: authService,
	}
}
